blenderproc/python/writer/GifWriterUtility.py

```diff
@@ -0,0 +1,210 @@
+    :param output_data_dict: The data dictionary which was produced
+                            by the render method.
+    :param append_to_existing_output: If this is True, the output_dir_path
+                            folder will be scanned for pre-existing
+                            files of the name #_animation.gif 
+                            and the number of newly added files will
+                            start right where the last run left off.
+    :param time_constant: Duration of each frame in the animation 
+                            in milliseconds.
+    :param time_direction: If this is True, the order of the frames
+                            will be reversed.
+    """
+
+    # Generates subdirectory for .gif files
+    output_dir_path = GifWriterUtility._provide_directory(output_dir_path)
+    
+    # From WriterUtility.py
+    amount_of_frames = 0
+    for data_block in output_data_dict.values():
+        if isinstance(data_block, list):
+            amount_of_frames = max([amount_of_frames, len(data_block)])
+
+    # From WriterUtility.py
+    if amount_of_frames != bpy.context.scene.frame_end - bpy.context.scene.frame_start:
+        raise Exception("The amount of images stored in the output_data_dict"
+                        "does not correspond with the amount"
+                        "of images specified by frame_start to frame_end.")
+
+    # Sorts out keys which are just metadata and not plottable
+    keys_to_use = GifWriterUtility._select_keys(output_data_dict)
+
+    # Build temporary folders with .png collections
+    to_animate = GifWriterUtility._cache_png(keys_to_use, output_data_dict)
+
+    # Write the cache .png files to .gif files and delete cache
+    GifWriterUtility._write_to_gif(to_animate, output_dir_path, 
+                                    append_to_existing_output, 
+                                    time_constant, 
+                                    time_direction)
+
+
+class GifWriterUtility:
+    
+    @staticmethod
+    def _provide_directory(output_dir_path: str) -> str:
+        """ Generates subdirectory for .gif files if not existent """
+        output_dir_path = os.path.join(output_dir_path, 'gif_animations')
+        if not os.path.exists(output_dir_path):
+            print('\n Generate output folder: ' + output_dir_path)
+            os.makedirs(output_dir_path)
+        return output_dir_path
+
+    @staticmethod
+    def _select_keys(
+            output_data_dict: Dict[str, List[Union[np.ndarray, list, dict]]]
+            ) -> List[str]:
+        """ Sorts out keys which are just metadata and not plottable """
+        keys_to_use = []
+        for key in output_data_dict.keys():
+            value = output_data_dict[key][0]
+            value = np.array(value)
+            if np.issubdtype(value.dtype, np.string_) or len(value.shape) == 1:
+                pass # metadata
+            else:
+                keys_to_use.append(key)
+        return keys_to_use
```

Comment:
```suggestion
    def _select_keys(
            output_data_dict: Dict[str, List[Union[np.ndarray, list, dict]]]
            ) -> List[str]:
        """ Sorts out keys which are just metadata and not plottable """
        def is_image(x: Union[np.ndarray, list, dict]) -> bool:
           """ Checks if the input x is not a string and is not a vector"""
            x = np.array(x)
            return not np.issubdtype(value.dtype, np.string_) and len(value.shape) != 1
       return [key for key, value in output_data_dict.items() if len(value) > 0 and is_image(value[0])]
```

This code is quite long and convoluted. You should not have empty if passes just negate the if statement then. Accessing the first element of a list without checking its length is also dangerous.